// nestjs-telegraf/dist/telegraf.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var TelegrafModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TelegrafModule = void 0;
const common_1 = require("@nestjs/common");
const telegraf_core_module_1 = require("./telegraf-core.module");
let TelegrafModule = TelegrafModule_1 = class TelegrafModule {
    static forRoot(options) {
        return {
            module: TelegrafModule_1,
            imports: [telegraf_core_module_1.TelegrafCoreModule.forRoot(options)],
            exports: [telegraf_core_module_1.TelegrafCoreModule],
        };
    }
    static forRootAsync(options) {
        return {
            module: TelegrafModule_1,
            imports: [telegraf_core_module_1.TelegrafCoreModule.forRootAsync(options)],
            exports: [telegraf_core_module_1.TelegrafCoreModule],
        };
    }
};
TelegrafModule = TelegrafModule_1 = __decorate([
    (0, common_1.Module)({})
], TelegrafModule);
exports.TelegrafModule = TelegrafModule;
